### PR TITLE
Test Information

### PR DESCRIPTION
**Pull request by m-ildefons**
Annotate the test information with Rancher/RKE version information. This helps the test report HTML files to be less ambiguous about the configuration under test.

### CODE REVIEW

File: apiclient/rancher_api/api.py [hunk 1]
```diff
@@ -56,8 +56,16 @@ def cluster_version(self):
...
             code, data = self.settings.get('server-version')
             ver = data['value']
-            # XXX: fix master-xxx-head to 8.8.8, need the API fix the problem
-            self._version = parse_version('8.8.8' if 'master' in ver else ver)
+
+            try:
+                # XXX: fix master-xxx-head to 8.8.8, need the API fix the problem
+                cur, _, _ = ver.split('-')
+                cur = "v8.8.8" if "master" in cur else cur
+                self._version = parse_version(f"{cur}.99")
```
Comment (albinsun) [inline apiclient/rancher_api/api.py:64]: Is this for branch head?
e.g. `v1.4-ba586eaf-head` -> `v1.4.99`
```
>>> ver = "v1.4.0"
>>> parse_version('8.8.8' if 'master' in ver else ver)
<Version('1.4.0')>    # OK
>>> 
>>> ver = "v1.4.0-rc1"
>>> parse_version('8.8.8' if 'master' in ver else ver)
<Version('1.4.0rc1')>    # OK
>>> 
>>> ver = "v1.4.0-dev-20240101"
>>> parse_version('8.8.8' if 'master' in ver else ver)
<Version('1.4.0.dev20240101')>    # OK
>>> 
>>> ver = "master"
>>> parse_version('8.8.8' if 'master' in ver else ver)
<Version('8.8.8')>    # OK
>>> 
>>> ver = "v1.4-ba586eaf-head"
>>> parse_version('8.8.8' if 'master' in ver else ver)
Traceback (most recent call last):
  ...
    raise InvalidVersion(f"Invalid version: '{version}'")
pkg_resources._vendor.packaging.version.InvalidVersion: Invalid version: 'v1.4-ba586eaf-head'
>>> 
>>> cur, _, _ = ver.split('-')
>>> cur = "v8.8.8" if "master" in cur else cur
>>> parse_version(f"{cur}.99")
<Version('1.4.99')>
```

File: apiclient/rancher_api/api.py [hunk 2]
```diff
@@ -56,8 +56,16 @@ def cluster_version(self):
...
+            except ValueError:
+                self._version = parse_version(ver)
```
Comment (albinsun) [inline apiclient/rancher_api/api.py:66]: It's better to keep current implementation in try and new one in exception.
```
try:
    self._version = parse_version('8.8.8' if 'master' in ver else ver)
except:
    #YOUR_NEW_IMPLEMENTATION
```

**Review by albinsun — COMMENTED** [review]
Rancher part LGTM. Have some comments on Harvester part.

File: apiclient/rancher_api/api.py (hunk 1 above)
Comment (m-ildefons) [reply apiclient/rancher_api/api.py]: I copied this code from the harvester API part assuming that Rancher follows a similar release naming. I updated it to match their release naming as best I could.

File: apiclient/rancher_api/api.py (hunk 2 above)
Comment (m-ildefons) [reply apiclient/rancher_api/api.py]: No, why?

**Comment by m-ildefons** [comment]
The goal of this PR is to enhance the metadata annotations of the test results and go from this:

Screenshot at 2024-09-17 10-04-48

to this:
Screenshot at 2024-09-17 10-05-02

File: apiclient/rancher_api/api.py [hunk 3]
```diff
@@ -56,8 +56,14 @@ def cluster_version(self):
         if not self._version:
             code, data = self.settings.get('server-version')
             ver = data['value']
-            # XXX: fix master-xxx-head to 8.8.8, need the API fix the problem
-            self._version = parse_version('8.8.8' if 'master' in ver else ver)
+
+            # Rancher versions look like v2.9.2, or v2.9.2-something
```
Comment (lanfon72) [inline apiclient/rancher_api/api.py:60]: Although Rancher is not using `master-head`, we would still need to consider cases like:
1. `v2.9.1-rc6`
2. `v2.9-2f9a7509687f559d772be75d61df7f7409803883-head`

As the second case would be the latest of `v2.9`, we need to fill the patch version for it. (to notice that there will not have something like `v2.9.1-sha-head`)

so, I would suggest that we could update it to something like
https://github.com/harvester/tests/blob/4d4d22b397e4fa8ff73ce193c6fce6b78624d356/apiclient/harvester_api/api.py#L43-L55

which we already implemented in `HarvesterAPI`.
To notice that Python usually use try-except to control the flow called EAFP.[^1]

[^1]: https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/

**Review by lanfon72 — CHANGES_REQUESTED** [review]

File: apiclient/rancher_api/api.py (hunk 3 above)
Comment (m-ildefons) [reply apiclient/rancher_api/api.py:60]: That was what I did initially, but @albinsun wasn't happy with it.

Comment (albinsun) [reply apiclient/rancher_api/api.py:60]: I was talking about the order but not drop error handling.
My thought was
1. Existing implementation works for most cases
   (Except head form like `v2.9-2f9a7509687f559d-head`)
2. I thought you are dealing the head form case, so it's natually to suggest
    ```
    try:
       # Existing implementation
    except:
       # Dealing exceptions
    ```

Anyway, per discussed, It's reasonable for me to align with Harvester API side implementation.

Comment (m-ildefons) [reply apiclient/rancher_api/api.py]: I think I have a good implementation now. It tries hard to parse the version, it keeps the suffix information for later reference and if all fails, it behaves like the Harvester API implementation.
I noticed that with my version of the `pkg_resources` library `parse_version` never raises an exception, but will produce a `LegacyVersion` object in case it doesn't know how to deal with it:
```
Python 3.6.15 (default, Sep 23 2021, 15:41:43) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pkg_resources import parse_version
>>> parse_version("foobar-noversion")
<LegacyVersion('foobar-noversion')>
```
Therefore I'm quite sure that the exception case will not raise another error.

Comment (albinsun) [reply apiclient/rancher_api/api.py]: As I know our daily env. adopts python 3.10 which has different behavior on `parse_version`
```
Python 3.10.14 (main, Mar 24 2024, 08:01:04) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pkg_resources import parse_version
<stdin>:1: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
>>> parse_version("foobar-noversion")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/albin/Workspace/albinsun/tests/.tox/py310/lib/python3.10/site-packages/pkg_resources/_vendor/packaging/version.py", line 200, in __init__
    raise InvalidVersion(f"Invalid version: '{version}'")
pkg_resources._vendor.packaging.version.InvalidVersion: Invalid version: 'foobar-noversion'
```

Comment (lanfon72) [reply apiclient/rancher_api/api.py]: `LegacyVersion` had been dropped in https://github.com/pypa/packaging/pull/407

File: apiclient/rancher_api/api.py [hunk 4]
```diff
@@ -55,9 +55,27 @@ def __init__(self, endpoint, token=None, session=None):
...
+            version = data['value']
+
+            # Rancher versions look like v2.9.2, or v2.9.2-suffix but also
+            # possibly v2.9-abcdef0123-head
+            try:
+                # Split off the version name
+                # parse the version name into a Version object if possible
+                # save the suffixes as one string, e.g.:
+                # v2.9.2               => <Version('2.9.2')>, suffix = None
+                # v2.9.2-rc1           => <Version('2.9.2')>, suffix = "rc1"
```
Comment (lanfon72) [inline apiclient/rancher_api/api.py:67]: To notice that the suffix is meaningful for the Version instance to compare with others, for example:
image

so we should only handle the case like `v2.9-2f9a7509687f559d772be75d61df7f7409803883-head`, as I know, rancher will not use `-dev-<date>` format, so it is fine to not consider the case.

**Review by albinsun — APPROVED** [review]
LGTM

**Comment by m-ildefons** [comment]
I've reduced this PR to the bare essential, which is to have the raw version string in the test metadata.

**Review by lanfon72 — APPROVED** [review]
